src/Ouroboros/Network/Socket.hs

```diff
@@ -0,0 +1,469 @@
+      x <- readLn
+      print ("Recv", x)
+      k x >>= go
+    go (Fail err) =
+      print ("Fail", err)
+
+demo1 :: IO ()
+demo1 = consoleProtocolAction example1
+
+-------------------------------------------
+
+-- | A demonstration that we can run the simple chain consumer protocol
+-- over a local socket with full message serialisation, framing etc.
+--
+demo2 :: (Chain.HasHeader block, Serialise block, Eq block)
+      => Chain block -> [ChainUpdate block] -> IO Bool
+demo2 chain0 updates = do
+
+    addr:_ <- getAddrInfo Nothing (Just "127.0.0.1") (Just "6060")
+    consSock <- socket (addrFamily addr) Stream defaultProtocol
+    setSocketOption consSock ReuseAddr 1
+    bind consSock (addrAddress addr)
+    listen consSock 2
+
+    prodSock <- socket (addrFamily addr) Stream defaultProtocol
+    connect prodSock (addrAddress addr)
+    (consSock', _) <- accept consSock
+
+    -- Initialise the producer and consumer state to be the same
+    producerVar <- newTVarIO (CPS.initChainProducerState chain0)
+    consumerVar <- newTVarIO chain0
+
+    -- Fork the producer and consumer
+    ptids <- runProducer prodSock (exampleProducer producerVar)
+    ctids <- runConsumer consSock' (exampleConsumer consumerVar)
+
+    -- Apply updates to the producer's chain and let them sync
+    _ <- forkIO $ sequence_
+           [ do threadDelay 10000 -- just to provide interest
+                atomically $ do
+                  p <- readTVar producerVar
+                  let Just p' = CPS.applyChainUpdate update p
+                  writeTVar producerVar p'
+           | update <- updates ]
+
+    -- Wait until the consumer's chain syncs with the producers chain
+    let Just expectedChain = Chain.applyChainUpdates updates chain0
+    chain' <- atomically $ do
+                chain' <- readTVar consumerVar
+                check (Chain.headPoint expectedChain == Chain.headPoint chain')
+                return chain'
+
+    --threadDelay 1500
+    {-sendAll prodSock $ pack "error"
+    sendAll consSock $ pack "error"
+    sendAll consSock' $ pack "error"-}
+
+    mapM_ cancel ptids
+    mapM_ cancel ctids
+    close prodSock
+    close consSock
+    close consSock'
+
+    return (expectedChain == chain')
+
+type ConsumerSideProtocol block = Protocol (MsgConsumer block) (MsgProducer block)
+type ProducerSideProtocol block = Protocol (MsgProducer block) (MsgConsumer block)
+type PongSideProtocol = Protocol MsgPong MsgPing
+type PingSideProtocol = Protocol MsgPing MsgPong
+
+data MsgPong = MsgPong deriving (Eq, Show)
+data MsgPing = MsgPing deriving (Eq, Show)
+
+instance Serialise MsgPong where
+  encode _ = encodeNull
+  decode = MsgPong <$ decodeNull
+
+instance Serialise MsgPing where
+  encode _ = encodeNull
+  decode = MsgPing <$ decodeNull
+
+
+
+runProducer :: forall block r. (Chain.HasHeader block, Serialise block)
+         => Socket -> ProducerHandlers block IO r -> IO [Async ()]
+runProducer sd producer = do
+    sem <- newQSem 0
```

Comment:
Why do we need to use semaphores? `STM` is not enough?

Reply:
The semaphore is used to to implement blocking read on multiple bounded queues. It is not possible to do that using `STM` alone.